clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/resources/appgateway/ops/UpsertAzureAppGatewayAtomicOperation.groovy

```diff
@@ -98,49 +99,75 @@ class UpsertAzureAppGatewayAtomicOperation implements AtomicOperation<Map> {
+          def vnetDescription = networkProvider.get(description.accountName, description.region, description.vnet)
 
-        if (!subnetName || vnet?.subnets?.find {it.name == subnetName}) {
-          // virtualNetworkName is not yet in the cache or the subnet we try to create already exists; we'll use the current vnet
-          //   we just got to re-compute the next subnet
-          vnetDescription = AzureVirtualNetworkDescription.getDescriptionForVirtualNetwork(vnet)
-          nextSubnetAddressPrefix = AzureVirtualNetworkDescription.getNextSubnetAddressPrefix(vnetDescription, rand.nextInt(vnetDescription?.maxSubnets ?: 1))
-          subnetName = AzureUtilities.getSubnetName(virtualNetworkName, nextSubnetAddressPrefix)
-        }
+          if (!vnetDescription) {
+            throw new RuntimeException("Selected virtual network $description.vnet does not exist")
+          }
 
-        task.updateStatus(BASE_PHASE, "Creating new subnet ${subnetName} for ${description.loadBalancerName}")
-        def subnetId = description.credentials.networkClient.createSubnet(resourceGroupName,
-          virtualNetworkName,
-          subnetName,
-          nextSubnetAddressPrefix,
-          description.securityGroup)
+          // subnet is valid only if it exists within the selected vnet and it's unassigned or all the associations are ALSO application gateways
+          subnetId = vnetDescription.subnets?.find { subnet ->
+            (subnet.name == description.subnet) && (!subnet.connectedDevices || !subnet.connectedDevices.find {it.type != "applicationGateways"})
+          }?.resourceId
 
-        if (!subnetId) {
-          task.updateStatus(BASE_PHASE, "Failed to create subnet for Application Gateway ${description.name}")
+          if (!subnetId) {
+            task.updateStatus(BASE_PHASE, "Failed to select subnet for Application Gateway ${description.name}")
+            throw new RuntimeException("Selected subnet $description.subnet in virtual network $description.vnet is not valid")
+          }
         } else {
-          description.vnet = virtualNetworkName
-          description.subnet = AzureUtilities.getNameFromResourceId(subnetId)
-
-          task.updateStatus(BASE_PHASE, "Create new application gateway ${description.loadBalancerName} in ${description.region}...")
-          DeploymentExtended deployment = description.credentials.resourceManagerClient.createResourceFromTemplate(description.credentials,
-            AzureAppGatewayResourceTemplate.getTemplate(description),
-            resourceGroupName,
-            description.region,
-            description.loadBalancerName,
-            "appGateway")
-
-          errList = AzureDeploymentOperation.checkDeploymentOperationStatus(task, BASE_PHASE, description.credentials, resourceGroupName, deployment.name)
-          loadBalancerName = description.name
+          task.updateStatus(BASE_PHASE, "Creating subnet for application gateway")
```

Comment:
Will add a new property on the description "useDefaultVnet", default to false, so that if user wants to automatically create a vnet/subnet, it must be intentional.